readthedocs/oauth/services/githubapp.py

```diff
@@ -0,0 +1,473 @@
+
+        Repositories that are no longer accessible to the installation are removed from the database
+        only if they are not linked to a project. This is in case the user wants to grant access to the repository again.
+
+        If a remote organization doesn't have any repositories after removing the repositories,
+        we remove the organization from the database.
+        """
+        try:
+            app_installation = self.app_installation
+        except GithubException as e:
+            log.info(
+                "Failed to get installation",
+                installation_id=self.installation.installation_id,
+                exc_info=True,
+            )
+            if e.status == 404:
+                # The app was uninstalled, we remove the installation from the database.
+                self.installation.delete()
+            raise SyncServiceError()
+
+        if app_installation.suspended_at is not None:
+            log.info(
+                "Installation is suspended",
+                installation_id=self.installation.installation_id,
+                suspended_at=app_installation.suspended_at,
+            )
+            # The installation is suspended, we don't have access to it anymore,
+            # so we just delete it from the database.
+            self.installation.delete()
+            raise SyncServiceError()
+
+        remote_repositories = []
+        for repo in app_installation.get_repos():
+            remote_repo = self._create_or_update_repository_from_gh(repo)
+            if remote_repo:
+                remote_repositories.append(remote_repo)
+
+        repos_to_delete = self.installation.repositories.exclude(
+            pk__in=[repo.pk for repo in remote_repositories],
+        ).values_list("remote_id", flat=True)
+        self.installation.delete_repositories(repos_to_delete)
+
+    def update_or_create_repositories(self, repository_ids: list[int]):
+        """Update or create repositories from the given list of repository IDs."""
+        for repository_id in repository_ids:
+            try:
+                # NOTE: we save the repository ID as a string in our database,
+                # in order for PyGithub to use the API to fetch the repository by ID (not by name).
+                # it needs to be an integer, so just in case we cast it to an integer.
+                repo = self.installation_client.get_repo(int(repository_id))
+            except GithubException as e:
+                log.info(
+                    "Failed to fetch repository from GitHub",
+                    repository_id=repository_id,
+                    exc_info=True,
+                )
+                # if we lost access to the repository,
+                # we remove the repository from the database,
+                # and clean up the collaborators and relations.
+                if e.status in [404, 403]:
+                    self.installation.delete_repositories([repository_id])
```

Comment:
Seems we could be calling this a bunch of individual times, instead of building a list of ids and deleting them at the end?